### PR TITLE
Adapter_Engine: Avoid signature duplication for AdapterId() methods

### LINKED ISSUE #284: Two methods AdapterId with the exact same name and list of parameters
#### Broken rules:

The two methods below have the exact same name name and list of parameters so things like `System.Type.GetMethod(string name, Type[] types)` for example doesn't work.

```
public static object AdapterId(this IBHoMObject bHoMObject, Type adapterIdFragmentType)
public static T AdapterId<T>(this IBHoMObject bHoMObject, Type adapterIdFragmentType)
```

### LINKED ISSUE #288: Improve the Id query
#### Description:

During today's workshop with the new users I got mindblown by how hard it is to extract the Id of an object in a given software. The workshop participants received a double blow when they needed to quickly learn what needs to be plugged in to `adapterIdFragmentType`, what a type is and what do fragments have to do with it 😃

IMO the UX of this is unacceptable, I did not even know that it works this way as I did create an `ElementId` query method in Revit_Toolkit a long time ago. Couldn't we simply have a set of methods like `RobotId`, `EtabsId`, `NameYourFavSoftwareId`? This will make it SO much easier for the users, who may not know what adapter is, but will for sure start the search of e.g. Robot Id from typing 'robot id' in the search bar.

image

### PR DESCRIPTION
**Pull request by alelom**
## Dependant PRs

- SAP2000 needs alignment to this PR, done in https://github.com/BHoM/SAP2000_Toolkit/pull/200.
- ETABS needs alignment to this PR, done in https://github.com/BHoM/ETABS_Toolkit/pull/380
   
### Issues addressed by this PR

Closes #284
Closes #288

This PR revises the two methods that get the `AdapterId` of a specific fragment Type out of a given BHoMObject.
Both methods are useful and need to be maintained, but due to the conflict explained in #284 they had to be renamed/revised.

We have:
1. a non generic method `AdapterId(this BHoMObject, ...)` that returns `object`. This gathers all IDs that implement a certain fragment type. If multiple IDs are found, a List is returned. If one Id is found, only that is returned.  
This method was renamed to `AdapterIds<T>(this BHoMObject, ...)` (note the **s**).
2. a generic method that does a similar thing, but also tries to cast to the requested type `<T>`. This method will return an error if multiple IDs are found. This method was slightly modified to expose more useful errors/warnings and better distinguish its function from the other one. The method was not renamed.

There are only 2 Toolkits that requires alignment (SAP2000/ETABS) because they have a reference to the renamed method (1).

All other Adapter Toolkits needs to be tested to make sure that the functionality is still working as expected.

### Test files

Needs testing of existing functionality on all Adapters.

### CODE REVIEW

**Comment by pawelbaran** [comment]
I will not approve not to give a false impression that this PR has been tested extensively, but can confirm that Revit_Toolkit works perfectly fine on this branch.

**Comment by pawelbaran** [comment]
Not sure whether it is fully relevant to this PR, but I have just raised #288 - being on this branch, I received a massive pushback from the users who were not able to extract any Ids without help. This tells me the UX will need further improvement once #284 is resolved.

**Comment by alelom** [comment]
> Not sure whether it is fully relevant to this PR, but I have just raised #288 - being on this branch, I received a massive pushback from the users who were not able to extract any Ids without help. This tells me the UX will need further improvement once #284 is resolved.

See [my reply there](https://github.com/BHoM/BHoM_Adapter/issues/288#issuecomment-781440780).

**Comment by IsakNaslundBh** [comment]
@BHoMBot check versioning

**Review by IsakNaslundBh — COMMENTED** [review]
Sorry for not reviewing... Been swamped!

Code changes generally look fine to me just looking here at github, but have not tested it yet.

Would assume this would require versioning, some PreviousVersion attributes to both of the updated methods!

**Comment by alelom** [comment]
> Would assume this would require versioning, some PreviousVersion attributes to both of the updated methods!

Thanks, actually only one method has been modified (AdapterId), the other (AdapterIds) is new. Will add versioning for it.

**Comment by IsakNaslundBh** [comment]
@BHoMBot check versioning

**Comment by IsakNaslundBh** [comment]
@BHoMBot check versioning

File: Adapter_Engine/Query/AdapterIds.cs [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        [Description("Returns the BHoMObject's Id of the provided FragmentType. " +
+            "If more than one matching IdFragment is found, the method returns a List of all Ids of that type." +
+            "If none is found, `null` is returned.")]
+        public static object AdapterIds(this IBHoMObject bHoMObject, Type adapterIdFragmentType = null)
```
Comment (IsakNaslundBh) [inline Adapter_Engine/Query/AdapterIds.cs:45]: Think this guy also would need a PrevVersion attribute? Cause it is a more or less straight renaming of the non-generic method, isnt it?

**Review by IsakNaslundBh — APPROVED** [review]
From what I can see, this is not causing any issues in GSA, Etabs or Robot.

I know it has also been tested for SAP2000.

@peterjamesnugent have just checked if this also works for Midas and Lusas?

File: Adapter_Engine/Query/AdapterId.cs [hunk 2]
```diff
@@ -39,43 +39,33 @@ public static partial class Query
...
+
                 return default(T);
             }
 
             if (id is T)
             {
                 return (T)id;
             }
+
+            if (id is IEnumerable)
```
Comment (IsakNaslundBh) [inline Adapter_Engine/Query/AdapterId.cs:63]: After debugging with @peterjamesnugent I think this change is needed

```suggestion
            if (id is IEnumerable && !(is id string))
```

**Review by IsakNaslundBh — CHANGES_REQUESTED** [review]
Failing for strings trying to convert to for example int. See below

File: Adapter_Engine/Query/AdapterId.cs (hunk 2 above)
Comment (alelom) [reply Adapter_Engine/Query/AdapterId.cs:63]: True, otherwise `string`s are intercepted. Good catch!

**Review by peterjamesnugent — CHANGES_REQUESTED** [review]
This works for Lusas, both Push/Pull for various objects.

I am running in to an issue when pushing to MidasCivil though on this line:
https://github.com/BHoM/BHoM_Adapter/blob/e80ed0bd7cca3414f7580ce84261e40e542ce2b9/BHoM_Adapter/CRUD/IRead.cs#L100

> System.MissingMethodException: 'Method not found: '!!0 BH.Engine.Adapter.Query.AdapterId(BH.oM.Base.IBHoMObject, System.Type)'.'

Basically it's the first read before it pushes (so should just be empty).

System.MissingMethodException
  HResult=0x80131513
  Message=Method not found: '!!0 BH.Engine.Adapter.Query.AdapterId(BH.oM.Base.IBHoMObject, System.Type)'.
  Source=MidasCivil_Adapter
  StackTrace:
   at BH.Adapter.MidasCivil.MidasCivilAdapter.ReadBars(List`1 ids)
   at BH.Adapter.MidasCivil.MidasCivilAdapter.IRead(Type type, IList ids, ActionConfig actionConfig)
   at BH.Adapter.BHoMAdapter.Read(Type type, String tag, ActionConfig actionConfig) in C:\Users\pnugent\Documents\GitHub\BHoM_Adapter\BHoM_Adapter\CRUD\IRead.cs:line 100

Script is here if you want to give it a try and debug:
https://burohappold.sharepoint.com/:f:/r/sites/BHoM/02_Current/12_Scripts/01_Issue/Archive/MidasCivil_Toolkit/MidasCivil_Toolkit-%23263/Elements?csf=1&web=1&e=erZTfB

**Review by peterjamesnugent — APPROVED** [review]
Tested for both Lusas and Midas, able to Push/Pull a number of different objects.

**Review by IsakNaslundBh — APPROVED** [review]
Works fine now!

Need to get https://github.com/BHoM/SAP2000_Toolkit/pull/209 merged first, then this should be good to go!

**Review by IsakNaslundBh — APPROVED** [review]
Re-approving

**Comment by alelom** [comment]
Replaced with https://github.com/BHoM/BHoM_Adapter/pull/290 due to:
image

**Pull request closed without merging**